types/coin.go

```diff
@@ -239,67 +307,111 @@ func (coins Coins) IsEqual(coinsB Coins) bool {
 	if len(coins) != len(coinsB) {
 		return false
 	}
+
+	sort.Sort(coins)
+	sort.Sort(coinsB)
+
 	for i := 0; i < len(coins); i++ {
 		if coins[i].Denom != coinsB[i].Denom || !coins[i].Amount.Equal(coinsB[i].Amount) {
 			return false
 		}
 	}
+
 	return true
 }
 
-// IsPositive returns true if there is at least one coin, and all
-// currencies have a positive value
-func (coins Coins) IsPositive() bool {
+// Empty returns true if there are no coins and false otherwise.
+func (coins Coins) Empty() bool {
+	return len(coins) == 0
+}
+
+// Returns the amount of a denom from coins
+func (coins Coins) AmountOf(denom string) Uint {
+	switch len(coins) {
+	case 0:
+		return ZeroUint()
+
+	case 1:
+		coin := coins[0]
+		if coin.Denom == denom {
+			return NewUintFromBigInt(coin.Amount.BigInt())
+		}
+		return ZeroUint()
+
+	default:
+		midIdx := len(coins) / 2 // 2:1, 3:1, 4:2
+		coin := coins[midIdx]
+
+		if denom < coin.Denom {
+			return coins[:midIdx].AmountOf(denom)
+		} else if denom == coin.Denom {
+			return NewUintFromBigInt(coin.Amount.BigInt())
+		} else {
+			return coins[midIdx+1:].AmountOf(denom)
+		}
+	}
+}
+
+// negative returns a set of coins with all amount negative.
+func (coins Coins) negative() Coins {
+	res := make([]Coin, 0, len(coins))
+
+	for _, coin := range coins {
+		res = append(res, Coin{
+			Denom:  coin.Denom,
+			Amount: coin.Amount.Neg(),
+		})
+	}
+
+	return res
+}
+
+// isPositive returns true if there is at least one coin and all currencies
+// have a positive value.
+func (coins Coins) isPositive() bool {
 	if len(coins) == 0 {
 		return false
 	}
+
 	for _, coin := range coins {
-		if !coin.IsPositive() {
+		if !coin.isPositive() {
 			return false
 		}
 	}
+
 	return true
 }
 
-// IsNotNegative returns true if there is no currency with a negative value
-// (even no coins is true here)
-func (coins Coins) IsNotNegative() bool {
+// isNotNegative returns true if there is no coin amount with a negative value
+// (even no coins is true here).
+func (coins Coins) isNotNegative() bool {
 	if len(coins) == 0 {
 		return true
 	}
+
 	for _, coin := range coins {
-		if !coin.IsNotNegative() {
+		if !coin.isNotNegative() {
 			return false
 		}
 	}
+
 	return true
 }
 
-// Returns the amount of a denom from coins
-func (coins Coins) AmountOf(denom string) Int {
-	switch len(coins) {
-	case 0:
-		return ZeroInt()
-	case 1:
-		coin := coins[0]
-		if coin.Denom == denom {
-			return coin.Amount
-		}
-		return ZeroInt()
-	default:
-		midIdx := len(coins) / 2 // 2:1, 3:1, 4:2
-		coin := coins[midIdx]
-		if denom < coin.Denom {
-			return coins[:midIdx].AmountOf(denom)
-		} else if denom == coin.Denom {
-			return coin.Amount
-		} else {
-			return coins[midIdx+1:].AmountOf(denom)
+// removeZeroCoins removes all zero coins from the given coin set.
+func removeZeroCoins(coins Coins) Coins {
+	var res Coins
```

Comment:
More copying :neutral_face: - can we just modify the array in place?

Reply:
Yes, of course. Good catch. Done!